### PR TITLE
mgmt: dfu: Necessary changes so DFU works with mcumgr.

### PR DESCRIPTION
**Pull request by mlaz**
- Added some Kconfig values so needed mcumgr.
- Added  boot_current_slot and made flash_sector_from_off visible so these would be used by mcumgr.

### CODE REVIEW

File: subsys/dfu/boot/mcuboot.c [hunk 1]
```diff
@@ -37,6 +37,9 @@
 #define BOOT_FLAG_UNSET 3
 #define BOOT_FLAG_ANY   4  /* NOTE: control only, not dependent on sector */
 
+/* Currently only used by imgmgr */
```
[the remark this answers is not in the document]
Comment (mlaz) [reply subsys/dfu/boot/mcuboot.c]: Changed.

**Comment by de-nordic** [comment]
I have tested this at 0c03364b8d6d33f7492cf68cfcaab1fe3a97b3aa, with mcumgr from branch  https://github.com/mlaz/mynewt-mcumgr/commit/9f5d349b100ff86a4ba1fb5af53e573cb85ed2c2 on default Kconfig settings for IMG_MGMT_ ; both progressive erase enabled and not enabled. Ble and serial transfer seem to work fine.

**Review by de-nordic — APPROVED** [review]

**Comment by de-nordic** [comment]
Is there anyone, from code-owners, with spare time to review this?

File: include/dfu/flash_img.h [hunk 2]
```diff
@@ -24,6 +24,18 @@ struct flash_img_context {
...
+/**
+ * @brief Get a flash sector which comprises a given offset.
+ *
+ * @param fap the flash area where the offset belongs
+ * @param off the offset
+ * @param sector the buffer to store the retrieved sector info
+ *
+ * @return 0 on success, negative errno code on fail.
+ */
+int flash_sector_from_off(struct flash_area const *fap, off_t off,
```
Comment (nvlsianpu) [inline include/dfu/flash_img.h]: pleas take https://github.com/zephyrproject-rtos/zephyr/pull/20973/ into account.
offset will be in relation to the fap beginning.

File: subsys/dfu/img_util/flash_img.c [hunk 3]
```diff
@@ -84,8 +83,10 @@ static int flash_sector_from_off(struct flash_area const *fap, off_t off,
 	}
 
 	return rc;
+
```
Comment (nvlsianpu) [inline subsys/dfu/img_util/flash_img.c]: unwanted newline

File: subsys/mgmt/Kconfig.mcumgr [hunk 4]
```diff
@@ -11,7 +11,12 @@ config MCUMGR
 
 if MCUMGR
 
-menu "Command handlers"
+menu "CBOR Attribute Library"
+config CBORATTR_MAX_SIZE
```
Comment (nvlsianpu) [inline subsys/mgmt/Kconfig.mcumgr]: would be great if key name will be in scope of mcumgr: MGMT_CBORATTR_MAX_SIZE etc

File: subsys/mgmt/Kconfig.mcumgr [hunk 5]
```diff
@@ -53,13 +58,33 @@ menuconfig MCUMGR_CMD_IMG_MGMT
...
-	config IMG_MGMT_UL_CHUNK_SIZE
+config IMG_MGMT_UL_CHUNK_SIZE
 	int "Maximum chunk size for image uploads"
 	default 512
 	help
 	  Limits the maximum chunk size for image uploads, in bytes.  A buffer of
 	  this size gets allocated on the stack during handling of a image upload
 	  command.
+
+config IMG_MGMT_LAZY_ERASE
```
Comment (nvlsianpu) [inline subsys/mgmt/Kconfig.mcumgr]: is this similar to [IMG_ERASE_PROGRESSIVELY](https://github.com/zephyrproject-rtos/zephyr/blob/master/subsys/dfu/Kconfig#L53)

**Comment by nvlsianpu** [comment]
@mlaz ^^ Just a few question and suggestions.

**Comment by mlaz** [comment]
@nvlsianpu thanks for reviewing, I'll address your suggestions shortly.

File: subsys/mgmt/Kconfig.mcumgr (hunk 5 above)
Comment (nvlsianpu) [reply subsys/mgmt/Kconfig.mcumgr]: check that it is similar. looks like not easy to use flash_util  in mcumgr now

Comment (mlaz) [reply subsys/mgmt/Kconfig.mcumgr]: It sure is, sorry for not stating it promptly. I will refactor in order to set IMG_MGMT_LAZY_ERASE with IMG_ERASE_PROGRESSIVELY.

**Comment by mlaz** [comment]
Taking into account the review on https://github.com/apache/mynewt-mcumgr/pull/46
changes were made so it now depends on https://github.com/zephyrproject-rtos/zephyr/pull/20973

**Comment by mlaz** [comment]
@nvlsianpu ^

**Review by nvlsianpu — COMMENTED** [review]
looks fine now, will just test it for being sure

**Comment by nvlsianpu** [comment]
last commit was just merged to the master

File: subsys/dfu/boot/mcuboot.c [hunk 6]
```diff
@@ -37,6 +37,12 @@
 #define BOOT_FLAG_UNSET 3
 #define BOOT_FLAG_ANY   4  /* NOTE: control only, not dependent on sector */
 
+/*
+ * The slot from which mcuboot boots from.
+ * Currently only used by imgmgr.
+ */
+int boot_current_slot;
```
Comment (utzig) [inline subsys/dfu/boot/mcuboot.c]: So, I think the original idea was that images could run on any slot, at least to be implemented some day; this was never done and maybe never will. This is could as well be removed and in the `mcumgr` repo references to to this variable could be changed to `0`.

**Comment by vrahane** [comment]
I think it is fine to remove `boot_current_slot`. Also, why is this marked DNM ? the config variables that changed are LOG specific config variables and the other ones are new variables which got added recently. I don't think any of this change is backwards incompatible particularly.

**Review by de-nordic — COMMENTED** [review]
_UPDATE: Since posting this comment I have approved this PR due to reasons specified below:
https://github.com/zephyrproject-rtos/zephyr/pull/20584#pullrequestreview-334635483_

@mlaz, @nvlsianpu Default settings compile and work fine with clean mcumgr, selected by west.
but...
**when I enable log backend for the mcumgr I am no longer able to compile the code**, even with your changes to the mcumgr (https://github.com/apache/mynewt-mcumgr/pull/46), because of error:
```
$$ west build -b nrf52840_pca10056 --build-dir smp_srv_zeph zephyr/samples/subsys/mgmt/mcumgr/smp_svr/
-- west build: build configuration:
       source directory: /home/doer/nordic-src/zephyrproject-1/zephyr/samples/subsys/mgmt/mcumgr/smp_svr
       build directory: /home/doer/nordic-src/zephyrproject-1/smp_srv_zeph
       BOARD: nrf52840_pca10056 (origin: CMakeCache.txt)
-- west build: building application
[0/1] Re-running CMake...
Zephyr version: 2.1.99
-- Selected BOARD nrf52840_pca10056
-- Found west: /home/doer/.local/bin/west (found suitable version "0.6.99", minimum required is "0.6.0")
-- Loading /home/doer/nordic-src/zephyrproject-1/zephyr/boards/arm/nrf52840_pca10056/nrf52840_pca10056.dts as base
Devicetree configuration written to /home/doer/nordic-src/zephyrproject-1/smp_srv_zeph/zephyr/include/generated/generated_dts_board.conf
Parsing Kconfig tree in /home/doer/nordic-src/zephyrproject-1/zephyr/Kconfig
Loaded configuration '/home/doer/nordic-src/zephyrproject-1/smp_srv_zeph/zephyr/.config'
No change to '/home/doer/nordic-src/zephyrproject-1/smp_srv_zeph/zephyr/.config'
-- Cache files will be written to: /home/doer/.cache/zephyr
CMake Error at /home/doer/nordic-src/zephyrproject-1/modules/lib/mcumgr/cmd/log_mgmt/CMakeLists.txt:12:
  Parse error.  Expected "(", got newline with text "

  ".


-- Configuring incomplete, errors occurred!
See also "/home/doer/nordic-src/zephyrproject-1/smp_srv_zeph/CMakeFiles/CMakeOutput.log".
See also "/home/doer/nordic-src/zephyrproject-1/smp_srv_zeph/CMakeFiles/CMakeError.log".
FAILED: build.ninja
/home/doer/.local/lib/python3.7/site-packages/cmake/data/bin/cmake -S/home/doer/nordic-src/zephyrproject-1/zephyr/samples/subsys/mgmt/mcumgr/smp_svr -B/home/doer/nordic-src/zephyrproject-1/smp_srv_zeph
ninja: error: rebuilding 'build.ninja': subcommand failed
ERROR: command exited with status 1: /home/doer/.local/bin/cmake --build /home/doer/nordic-src/zephyrproject-1/smp_srv_zeph
$$=>1 | doer@local-bm:~/nordic-src/zephyrproject-1
$$
```

**I have also have found that log backend for mcumgr does not seem to be compiling at latest zephyr/master**, witch missing entire includes (`mdlog.h`) and definitions of `mdlog_*` named functions that are required for this backend.

@mlaz: Does your mcumgr PR (https://github.com/apache/mynewt-mcumgr/pull/46) fix the loggng? If yes then this PR should be rather merged AFTER the mcumgr and should contain update to west.yml to point the changes.

Otherwise we (@nvlsianpu, @mlaz, @de-nordic ) should discuss temporarily removing the non-compiling option from Kconfig, until correct solution is provided.

**Review by de-nordic — APPROVED** [review]
Since the log backend is broken on master branch, I do not think that it should be fixed with this PR, separate PR should be created to provide the fix. This PR should allow us to bring DFU up to date.

I think that we can pass this PR through as the default compilation works fine.

Do you agree @nvlsianpu ?

File: subsys/dfu/boot/mcuboot.c (hunk 6 above)
Comment (nvlsianpu) [reply subsys/dfu/boot/mcuboot.c]: ^^ If changing, better to a macro defined value. Actually I have request for any image boot in my backlog.

**Review by nvlsianpu — APPROVED** [review]
I'm in line with @de-nordic : what doesn't work on master can be fixed in a following PR

File: subsys/dfu/boot/mcuboot.c (hunk 6 above)
Comment (mlaz) [reply subsys/dfu/boot/mcuboot.c]: `boot_current_slot` was now removed and a macro was defined in order to be used on img_mgmt.c

**Comment by mlaz** [comment]
west.yml now points to https://github.com/zephyrproject-rtos/mcumgr/pull/11 as requested by @de-nordic

**Comment by de-nordic** [comment]
@mlaz please fix Gitlint  issues

**Comment by mlaz** [comment]
@de-nordic fixed and rebased to master.

File: subsys/mgmt/Kconfig.mcumgr [hunk 7]
```diff
@@ -10,8 +10,13 @@ config MCUMGR
 
 if MCUMGR
 
-menu "Command handlers"
+config MGMT_CBORATTR_MAX_SIZE
+  int "The maximum size of a CBOR attribute during decoding"
+	default 512
+	help
```
Comment (jhedberg) [inline subsys/mgmt/Kconfig.mcumgr]: The indentation seems to be off here for the last two lines

File: subsys/mgmt/Kconfig.mcumgr [hunk 8]
```diff
@@ -89,8 +108,15 @@ config LOG_MGMT_BODY_LEN
 	help
 	  Limits the maximum length of log entry bodies, in bytes.  If a log
 	  entry's body length exceeds this number, it gets truncated in management
-	  responses.  A buffer of this size gets allocated on the stack during
+	  responses. A buffer of this size gets allocated on the stack during
 	  handling of the log show command.
+
+config LOG_READ_WATERMARK_UPDATE
+	bool "Enable reading of log watermark update"
+	default n
```
Comment (jhedberg) [inline subsys/mgmt/Kconfig.mcumgr]: `n` is the implicit default so you can remove this line

File: subsys/mgmt/Kconfig.mcumgr [hunk 9]
```diff
@@ -52,13 +57,26 @@ menuconfig MCUMGR_CMD_IMG_MGMT
...
 	int "Maximum chunk size for image uploads"
 	default 512
 	help
 	  Limits the maximum chunk size for image uploads, in bytes.  A buffer of
 	  this size gets allocated on the stack during handling of a image upload
 	  command.
+
+config IMG_MGMT_VERBOSE_ERR
+	bool "Verbose logging when uploading a new image"
+	default n
```
Comment (jhedberg) [inline subsys/mgmt/Kconfig.mcumgr]: `n` is the implicit default, so this line is redundant

File: subsys/mgmt/Kconfig.mcumgr [hunk 10]
```diff
@@ -52,13 +57,26 @@ menuconfig MCUMGR_CMD_IMG_MGMT
...
+	help
+	  Enable verbose logging during a firmware upgrade.
+
+config IMG_MGMT_DUMMY_HDR
+	bool "Return dummy image header data for imgr functions"
+	default n
```
Comment (jhedberg) [inline subsys/mgmt/Kconfig.mcumgr]: same here

File: subsys/mgmt/Kconfig.mcumgr (hunk 9 above)
Comment (mlaz) [reply subsys/mgmt/Kconfig.mcumgr]: fixed

File: subsys/mgmt/Kconfig.mcumgr (hunk 7 above)
Comment (mlaz) [reply subsys/mgmt/Kconfig.mcumgr]: fixed

**Comment by mlaz** [comment]
@jhedberg just pushed the requested changes

**Review by carlescufi — APPROVED** [review]
LGTM

**Comment by carlescufi** [comment]
@mlaz please rebase

**Comment by mlaz** [comment]
@carlescufi rebased

**Comment by carlescufi** [comment]
@mlaz thanks. To be merged once the 2.3 window opens.

**Pull request merged**